vulnfeeds/cpp/main.go

```diff
@@ -0,0 +1,238 @@
+	scanner.Scan()
+	if scanner.Text() != "Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/" {
+		return "", false
+	}
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		match := re.FindStringSubmatch(line)
+		if match != nil {
+			return match[1], true
+		}
+	}
+
+	return "", false
+}
+
+// Tries to find a Debian copyright file for the package and returns the source URL if IsRepoUrl() agrees.
+func MaybeGetSourceRepoFromDebian(mdir string, pkg string) string {
+	var metadata string
+	if strings.HasPrefix(pkg, "lib") {
+		metadata = path.Join(mdir, "changelogs/main", string(pkg[0:4]), pkg, "unstable_copyright")
+	} else {
+		metadata = path.Join(mdir, "changelogs/main", string(pkg[0]), pkg, "unstable_copyright")
+	}
+	if _, err := os.Stat(metadata); err == nil {
+		// parse the copyright file and go from here
+		Logger.Infof("FYI: Will look at %s", metadata)
+		possibleRepo, ok := MaybeGetSourceFromDebianCopyright(metadata)
+		if !ok {
+			return ""
+		}
+		if IsRepoUrl(possibleRepo) {
+			return possibleRepo
+		}
+		// Incorporate Savannah URL to Git translation here
+		if IsGnuUrl(possibleRepo) {
+			repo, translated := MaybeTranslateSavannahURL(possibleRepo)
+			if translated {
+				return repo
+			}
+		}
+		Logger.Infof("FYI: Disregarding %s", possibleRepo)
+	}
+	return ""
+}
+
+func main() {
+	jsonPath := flag.String("nvd_json", "", "Path to NVD CVE JSON to examine.")
+	debianMetadataPath := flag.String("debian_metadata_path", "", "Path to Debian copyright metadata")
+
+	flag.Parse()
+
+	client, err := logging.NewClient(context.Background(), projectId)
+	if err != nil {
+		log.Fatalf("Failed to create client: %v", err)
+	}
+	defer client.Close()
+
+	Logger.GCloudLogger = client.Logger("cpp-osv")
+
+	data, err := ioutil.ReadFile(*jsonPath)
+	if err != nil {
+		Logger.Fatalf("Failed to open file: %v", err) // double check this is best practice output
+	}
+
+	var parsed cves.NVDCVE
+	err = json.Unmarshal(data, &parsed)
+	if err != nil {
+		Logger.Fatalf("Failed to parse NVD CVE JSON: %v", err)
+	}
+
+	for _, cve := range parsed.CVEItems {
+		refs := cve.CVE.References.ReferenceData
+		patch_refs := 0
+		cpes := cves.CPEs(cve)
+
+		if len(refs) == 0 && len(cpes) == 0 {
+			Logger.Infof("FYI: skipping %s due to:", cve.CVE.CVEDataMeta.ID)
+			Logger.Infof("\t * lack of CPEs")
+			Logger.Infof("\t * lack of references")
+			continue
+		}
+
+		// Does it have any application CPEs?
+		app_cpes := 0
+		for _, cpe_str := range cves.CPEs(cve) {
+			cpe, ok := cves.ParseCPE(cpe_str)
+			if ok {
+				if cpe.Part == "a" {
+					app_cpes += 1
+				}
+			} else {
+				Logger.Fatalf("Failed to parse CPE %s: %v", cpe_str, err)
+			}
+		}
+
+		if app_cpes == 0 {
+			Logger.Infof("FYI: skipping %s due to:", cve.CVE.CVEDataMeta.ID)
+			Logger.Infof("\t * believed non-software")
+			continue
+		}
+
+		Logger.Infof("%s", cve.CVE.CVEDataMeta.ID)
+		for _, ref := range refs {
+			// Are any of the reference's tags 'Patch'?
```

Comment:
How accurate is this tagging? Do we lose (or gain) much if we always just use `IsRepoURL` to check these instead of relying on the tag?

Reply:
We gain a lot of noise (there's plenty of for example GitHub issue URLs that match when we stop filtering on patch tag).
It's more difficult at the moment to see if we wind up losing out. I've added a TODO to quantify this when the code evolves to actually deal with a data structure for each CVE.

Reply:
We can probably parse and detect the ones that are commits and pull requests pretty easily. Ack on TODO.